Enable local_partial_types

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118491
* #118481
* #118480
* #118479
* #118475
* #118469
* #118468
* __->__ #118467
* #118432
* #118418
* #118414

When using dmypy, this setting is enabled and cannot be turned off. Force it for regular mypy too.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

ezyang committed:
Enable local_partial_types

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

Skylion007 commented on test/test_torch.py:
```diff
@@ -10053,7 +10053,7 @@ class MyTwoTensor4(TwoTensor):
 INPLACE_METHOD = 2
 FUNCTIONAL = 4
-DIM_ARG = None
+DIM_ARG: None = None
```
Huh, never seen this annotation before. Didn't know that worked (should you use it instead of `type(None)`?)

Skylion007 commented on torch/_functorch/_aot_autograd/utils.py:
```diff
@@ -117,10 +117,10 @@ def call_func_at_runtime_with_args(f, args, steal_args=False, disable_amp=False)
-        None  # if the output spec is a tuple/list, we won't bother unflattening it.
-    )
-    is_really_simple = None  # if the output spec is a LeafSpec
+    is_simple: Optional[
```
I have a ruff rule for eradicating implicit optionals, but I couldn't figure out how to get all the backwards compatibility tests passing. `RUF013` if you want to give it a try. Also has autofixes.

ezyang replied:
For what it's worth, implicit optionals isn't really a priority for me right now, I want to target bigger semantic confusion in the codebase first lol

Skylion007 commented on torch/utils/data/datapipes/dataframe/dataframes.py:
```diff
@@ -416,7 +416,7 @@ def __getattr__(self, attrname):  # ?
 @functional_datapipe('trace_as_dataframe')
 class DataFrameTracer(CaptureDataFrameWithDataPipeOps, IterDataPipe):  # type: ignore[misc]
-    source_datapipe = None
+    source_datapipe: Optional[Any] = None
```
equivalent to ```suggestion
    source_datapipe: Optional = None
```
right?

Skylion007 commented on torch/utils/data/datapipes/dataframe/dataframe_wrapper.py:
```diff
@@ -1,5 +1,7 @@
-_WITH_PANDAS = None
+from typing import Any, Optional
+
+_pandas: Any = None
```
```suggestion
_pandas: Optional = None
```

ezyang replied:
So this I *know* will flag a rule somewhere saying that you shouldn't omit type parameters from generics...

ezyang replied:
https://mypy.readthedocs.io/en/stable/config_file.html#confval-disallow_any_generics

ezyang replied:
Oh this triggers for Type too

```
$ mypy --disallow-any-generics a.py
a.py:3: error: Missing type parameters for generic type "Type"  [type-arg]
Found 1 error in 1 file (checked 1 source file)
```

Skylion007 replied:
Ah, makes sense. Cary on then.

Skylion007 commented on torch/jit/_state.py:
```diff
@@ -60,8 +61,8 @@ def enable():
 # python class => ScriptClass mapping
-_script_classes = {}
-_name_to_pyclass = {}
+_script_classes: Dict[Type[Any], Type[Any]] = {}
+_name_to_pyclass: Dict[str, Type[Any]] = {}
```
```suggestion
_script_classes: Dict[Type[Any], Type] = {}
_name_to_pyclass: Dict[str, Type] = {}
```

Skylion007 reviewed (approved):
Some nits about unnecessary use of Any

Skylion007 commented on torch/serialization.py:
```diff
@@ -67,7 +67,7 @@ def mkdtemp():
 
 
-_package_registry = []
+_package_registry: List[Tuple[int, Callable[[STORAGE], Optional[str]], Callable[[STORAGE, str], Optional[STORAGE]]]] = []
```
Maybe want to separate out this tuple into a private TypeAlias.

ezyang committed:
Update on "Enable local_partial_types"


When using dmypy, this setting is enabled and cannot be turned off. Force it for regular mypy too.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Enable local_partial_types"


When using dmypy, this setting is enabled and cannot be turned off. Force it for regular mypy too.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Enable local_partial_types"


When using dmypy, this setting is enabled and cannot be turned off. Force it for regular mypy too.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Enable local_partial_types"


When using dmypy, this setting is enabled and cannot be turned off. Force it for regular mypy too.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Enable local_partial_types"


When using dmypy, this setting is enabled and cannot be turned off. Force it for regular mypy too.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
